jsre/natto.go

```diff
@@ -0,0 +1,170 @@
+
+// The VM has the following functions available:
+//
+//      <timer> = setTimeout(<function>, <delay>, [<arguments...>])
+//      <timer> = setInterval(<function>, <delay>, [<arguments...>])
+//      clearTimeout(<timer>)
+//      clearInterval(<timer>)
+//
+
+*/
+package jsre
+
+import (
+	"github.com/robertkrimen/otto"
+	"time"
+)
+
+type nattoTimer struct {
+	timer    *time.Timer
+	duration time.Duration
+	interval bool
+	call     otto.FunctionCall
+}
+
+type evalReq struct {
+	code   interface{}
+	result otto.Value
+	err    error
+	done   chan bool
+}
+
+type natto struct {
+	vm      *otto.Otto
+	eval    chan *evalReq
+	stop    chan bool
+	running bool
+}
+
+func newNatto(vm *otto.Otto) (self *natto) {
+
+	self = &natto{
+		vm:   vm,
+		eval: make(chan *evalReq),
+		stop: make(chan bool),
+	}
+	return
+}
+
+func (self *natto) nattoEventLoop() {
+
+	registry := map[*nattoTimer]*nattoTimer{}
+	ready := make(chan *nattoTimer)
+
+	newTimer := func(call otto.FunctionCall, interval bool) (*nattoTimer, otto.Value) {
+
+		delay, _ := call.Argument(1).ToInteger()
+		if 0 >= delay {
+			delay = 1
+		}
+		timer := &nattoTimer{
+			duration: time.Duration(delay) * time.Millisecond,
+			call:     call,
+			interval: interval,
+		}
+		registry[timer] = timer
+
+		timer.timer = time.AfterFunc(timer.duration, func() {
+			ready <- timer
+		})
+
+		value, err := call.Otto.ToValue(timer)
+		if err != nil {
+			panic(err)
+		}
+
+		return timer, value
+	}
+
+	setTimeout := func(call otto.FunctionCall) otto.Value {
+		_, value := newTimer(call, false)
+		return value
+	}
+
+	setInterval := func(call otto.FunctionCall) otto.Value {
+		_, value := newTimer(call, true)
+		return value
+	}
+
+	clearTimeout := func(call otto.FunctionCall) otto.Value {
+		timer, _ := call.Argument(0).Export()
+		if timer, ok := timer.(*nattoTimer); ok {
+			timer.timer.Stop()
+			delete(registry, timer)
+		}
+		return otto.UndefinedValue()
+	}
+
+loop:
+	for {
+		select {
+		case timer := <-ready:
+			// execute callback, remove/reschedule the timer
+			var arguments []interface{}
+			if len(timer.call.ArgumentList) > 2 {
+				tmp := timer.call.ArgumentList[2:]
+				arguments = make([]interface{}, 2+len(tmp))
+				for i, value := range tmp {
+					arguments[i+2] = value
+				}
+			} else {
+				arguments = make([]interface{}, 1)
+			}
+			arguments[0] = timer.call.ArgumentList[0]
+			_, err := self.vm.Call(`Function.call.call`, nil, arguments...)
+
+			if err != nil {
+				for _, timer := range registry {
+					timer.timer.Stop()
+					delete(registry, timer)
+					return
+				}
+			}
+			if timer.interval {
+				timer.timer.Reset(timer.duration)
+			} else {
+				delete(registry, timer)
+			}
+		case evalReq := <-self.eval:
+			// run the code, send the result back
+			self.vm.Set("setTimeout", setTimeout)
+			self.vm.Set("setInterval", setInterval)
+			self.vm.Set("clearTimeout", clearTimeout)
+			self.vm.Set("clearInterval", clearTimeout)
```

Comment:
When you move the logic into `JSRE`, you can register these during initialization.